### ISSUE TITLE
[BUG]: gradient accumulate not work for the first gradient_accumulation iters.

### ISSUE THREAD

**Issue by powermano** [hpcaitech/ColossalAI#1035]
### 🐛 Describe the bug

my config is 
```
from colossalai.amp import AMP_TYPE

BATCH_SIZE = 64
NUM_EPOCHS = 1
LOGGING_FREQUNCE = 100
OUTPUT = './'

# CONFIG = dict(fp16=dict(mode=AMP_TYPE.TORCH))
fp16 = dict(
    mode=AMP_TYPE.TORCH,
)


gradient_accumulation = 4

gradient_clipping = 5.0
```

My  main code is 
```

model = resnet34(num_classes=10)

# build dataloader
train_dataloader = get_dataloader(
        DATA_ROOT, local_rank=local_rank, batch_size=gpc.config.BATCH_SIZE, dali=dali, aug=aug)


# build criterion
criterion = torch.nn.CrossEntropyLoss()

# optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9, weight_decay=5e-4)

# lr_scheduler
lr_scheduler = CosineAnnealingLR(optimizer, total_steps=gpc.config.NUM_EPOCHS)

engine, train_dataloader, _, _ = colossalai.initialize(
    model,
    optimizer,
    criterion,
    train_dataloader,
)


ele_1st  = []
for epoch in range(gpc.config.NUM_EPOCHS):
        engine.train()
        for idx, (img, label) in enumerate(train_dataloader):
            img = img.cuda()
            label = label.cuda()

            engine.zero_grad()
            output, _ = engine(img, label)
            train_loss = engine.criterion(output, label)
            engine.backward(train_loss)
            engine.step()   

            ele_1st = next(model.parameters()).flatten()[-1]
            param_by_iter.append(ele_1st.item())
            
            # only run for 4 iterations
            if idx == 11:
                break

        lr_scheduler.step()
        train_dataloader.reset()
        
        for iteration, val in enumerate(param_by_iter):
            print(f'iteration {iteration} - value: {val}')

        if param_by_iter[-1] != param_by_iter[0]:
            print('The parameter is only updated in the last iteration')

```

And the result is 
```
iteration 0 - value: 0.026410149410367012
iteration 1 - value: 0.026410149410367012
iteration 2 - value: 0.026410149410367012
iteration 3 - value: 0.026410149410367012
iteration 4 - value: 0.026410149410367012
iteration 5 - value: 0.026410149410367012
iteration 6 - value: 0.026410149410367012
iteration 7 - value: 0.026409029960632324
iteration 8 - value: 0.026409029960632324
iteration 9 - value: 0.026409029960632324
iteration 10 - value: 0.026409029960632324
iteration 11 - value: 0.026399604976177216
The parameter is only updated in the last iteration
```

But when i try the tutorial of gradient accmulate, it actually works.  But i can not find the problems.

### Environment

pip install colossalai==0.1.4+torch1.10cu11.1 -f https://release.colossalai.org
ubuntu 20.04

**Comment by FrankLeeeee** [comment]
If you wish to verify this, you can turn off the torch AMP. In AMP, the first few steps may get skipped.

**Comment by powermano** [comment]
> If you wish to verify this, you can turn off the torch AMP. In AMP, the first few steps may get skipped.

thanks,

**Comment by powermano** [comment]
> 

By the way,  If I only have 4 GPUs, can pipeline parallelism be much faster than normal pytorch data parallelism? (When using pytorch's data parallel, GPU utilization is at 99%.)

**Comment by powermano** [comment]
> If you wish to verify this, you can turn off the torch AMP. In AMP, the first few steps may get skipped.

By the way, If I only have 4 GPUs, can pipeline parallelism be much faster than normal pytorch data parallelism? (When using pytorch's data parallel, GPU utilization is at 99%.)

**Comment by FrankLeeeee** [comment]
If your model can be accommodated on a single GPU, than data parallelism is usually  what you want.

**Comment by FrankLeeeee** [comment]
You can also use zero to boost your data parallel training and it can bring extra speedup.

**Comment by powermano** [comment]
> You can also use zero to boost your data parallel training and it can bring extra speedup.

I got it, thanks for your professional answer。
